### PR TITLE
features/highwayhash

### PR DESCRIPTION
**Pull request by aholler2**
We are building tensorflow for spack for HackIllinois. Highwayhash is one of the dependencies for tensorflow. It was long and painful and apparently simple and I want to sleep and I don't like building things.

### CODE REVIEW

File: var/spack/repos/builtin/packages/highwayhash/package.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+class Highwayhash(MakefilePackage):
+    """Strong (well-distributed and unpredictable) hashes:
+        - Portable implementation of SipHash
+        - HighwayHash, a 5x faster SIMD hash with security claims
+    """
+
+    homepage = "https://github.com/google/highwayhash"
+
+    version('dfcb97', git='https://github.com/google/highwayhash.git',
```
Comment (davydden) [inline var/spack/repos/builtin/packages/highwayhash/package.py]: Please name it 'develop'

File: var/spack/repos/builtin/packages/highwayhash/package.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            commit='dfcb97ca4fe9277bf9dc1802dd979b071896453b')
+
+    build_targets = ['all', 'libhighwayhash.a']
+
+    def install(self, spec, prefix):
+        install('libhighwayhash.a', prefix.lib)
+        for i in glob('highwayhash/*.h'):
+            install(i, prefix.bin)
```
Comment (davydden) [inline var/spack/repos/builtin/packages/highwayhash/package.py]: headers are installed in `prefix.bin`?!

Comment (tgamblin) [inline var/spack/repos/builtin/packages/highwayhash/package.py]: @aholler2: can you install the headers in prefix.include and the binaries made by the `all` target in prefix.bin?

File: var/spack/repos/builtin/packages/highwayhash/package.py (hunk 1 above)
Comment (citibeth) [inline var/spack/repos/builtin/packages/highwayhash/package.py]: Please name it `0.0.0.1`.  For security purposes, packages need at least one download that can be verified --- either by checksum, or by Git commit ID.  You can also add a "latest" version based on the `develop` branch, if you like.

File: var/spack/repos/builtin/packages/highwayhash/package.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    homepage = "https://github.com/google/highwayhash"
+
+    version('dfcb97', git='https://github.com/google/highwayhash.git',
+            commit='dfcb97ca4fe9277bf9dc1802dd979b071896453b')
+
+    build_targets = ['all', 'libhighwayhash.a']
+
+    def install(self, spec, prefix):
+        install('libhighwayhash.a', prefix.lib)
```
Comment (citibeth) [inline var/spack/repos/builtin/packages/highwayhash/package.py]: Can Highway Hash do shared libraries?

**Comment by tgamblin** [comment]
@citibeth: I think these might be easier to maintain if we just use the commit ids, as otherwise we have to cross-reference the package every time we update TF. 

What if instead we always mark the latest commit as preferred?

I also think it might be useful to revisit the versioning discussion later this year as I think it would help this and the environments working group... some of your ideas about ways to sort arbitrary versions apply here, although I haven't had time to completely think them through.

Also, it occurs to me that ideally we wouldn't need to write out explicit git versions.  We could just say a package has a git repo and other packages can depend on particular commits.

**Comment by citibeth** [comment]
> What if instead we always mark the latest commit as preferred?
>
That would work, I suppose.  But it would be impossible to compare version
numbers in `where=` clasues, or conditional logic inside of `install()`.

Thinking it over again, the original idea to just call this `develop` isn't
so bad.  As long as it's a specific git hash, then this version would still
be checksummable.  However, I believe (but am not quite sure) that we might
have logic that assumes that `develop` is a non-checksummable version.

The right approach here is to overhaul Spack's security model.  Something
I've thought about, but haven't done obviously.

I also think it might be useful to revisit the versioning discussion later
> this year as I think it would help this and the environments working
> group... some of your ideas about ways to sort arbitrary versions apply
> here, although I haven't had time to completely think them through.
>

> @citibeth <https://github.com/citibeth>: I think these might be easier to
> maintain if we just use the commit ids, as otherwise we have to
> cross-reference the package every time we update TF.
>
Also, it occurs to me that ideally we wouldn't need to write out explicit
> git versions. We could just say a package has a git repo and other packages
> can depend on particular commits.
>

What is TF?  I think this is the right long-term approach.  There is no
intrinsic need to manually specify every version that there is, as long as
Spack has other ways of knowing/finding versions.

The first step on this path is pluggable Version schemes #2179

A "poor man's" approach would be to create a version (say, `tf`) tied to a
specific Git hash.  We change the git hash associated with `tf` as needed
when we change the `tf` package.  But really, this seems to be not the way
Spack was intended.

**Comment by aholler2** [comment]
There are just a couple of things from highwayhash's Makefile that concern me. The following is a section from it. 
````
# CPU-specific implementations (same source file, different compiler flags)

libhighwayhash.a: $(SIP_OBJS) $(HIGHWAYHASH_OBJS)
	$(AR) rcs $@ $^

highwayhash_test: $(HIGHWAYHASH_OBJS) $(HIGHWAYHASH_TEST_OBJS)
	$(CXX) $(CXXFLAGS) $^ -o $@

benchmark: $(SIP_OBJS) $(HIGHWAYHASH_OBJS) $(BENCHMARK_OBJS)
	$(CXX) $(CXXFLAGS) $^ -o $@
````

I install these files regardless without any checks in the package. Would there be a way to check to see if these would be made when running make? I ask because if they weren't created, the install functions for them would break the install process, right? If you want to look at the Makefile to see it in context here is the link for it

[https://github.com/google/highwayhash/blob/dfcb97ca4fe9277bf9dc1802dd979b071896453b/Makefile](url)

**Comment by adamjstewart** [comment]
> Would there be a way to check to see if these would be made when running make?

```python
import os
os.path.exists('filename')
```

**Comment by citibeth** [comment]
Would it be possible for someone to just make a real CMake build for this
package upstream?  It looks like a pretty simple package.  Creating a CMake
build might be easier than all the gymnastics we would have to go through
to deal with a one-off manual Makefile.

**Comment by becker33** [comment]
@citibeth Can you approve this since you've retracted the opinion that the name should be '0.0.0.1'?

**Comment by tgamblin** [comment]
Thanks @aholler2!

**Comment by tgamblin** [comment]
Thanks @adamjstewart, @davydden, and @citibeth for all the feedback.  I've been busy lately but the project chugs along thanks to you guys.

**Pull request merged**